File: worker.js
/**
 * Cloudflare Workers entry point for Discord bot using Interactions API.
 * This allows the bot to run on Cloudflare Workers as a webhook-based application.
 */

import { verifyKey } from 'discord-interactions';

const COMMAND_HANDLERS = {
  ping: handlePingCommand,
  login: handleLoginCommand,
  'start-shift': handleStartShiftCommand,
  'end-shift': handleEndShiftCommand,
  'shift-status': handleShiftStatusCommand,
};

/**
 * Main request handler for Cloudflare Workers
 */
export default {
  async fetch(request, env) {
    if (request.method !== 'POST') {
      return new Response('Method not allowed', { status: 405 });
    }

    const signature = request.headers.get('X-Signature-Ed25519');
    const timestamp = request.headers.get('X-Signature-Timestamp');

    if (!signature || !timestamp) {
      return new Response('Missing signature headers', { status: 401 });
    }

    const body = await request.text();

    // Verify the request signature
    const isValidRequest = verifyKey(
      body,
      signature,
      timestamp,
      env.DISCORD_PUBLIC_KEY
    );

    if (!isValidRequest) {
      return new Response('Invalid signature', { status: 401 });
    }

    const interaction = JSON.parse(body);

    // Handle PING (Discord verification)
    if (interaction.type === 1) {
      return jsonResponse({ type: 1 });
    }

    // Handle APPLICATION_COMMAND
    if (interaction.type === 2) {
      const commandName = interaction.data.name;
      const handler = COMMAND_HANDLERS[commandName];

      if (handler) {
        return handler(interaction, env);
      }

      return jsonResponse({
        type: 4,
        data: {
          content: `Unknown command: ${commandName}`,
          flags: 64, // Ephemeral
        },
      });
    }

    return jsonResponse({
      type: 4,
      data: {
        content: 'Unsupported interaction type',
        flags: 64,
      },
    });
  },
};

/**
 * Handle ping command
 */
function handlePingCommand(interaction, env) {
  return jsonResponse({
    type: 4,
    data: {
      embeds: [
        {
          title: '🏓 Pong!',
          description: 'Bot is running on Cloudflare Workers!',
          color: 0xbebefe,
        },
      ],
    },
  });
}

/**
 * Handle login command - processes auth key linking
 */
async function handleLoginCommand(interaction, env) {
  const authKey = interaction.data.options?.find(opt => opt.name === 'auth_key')?.value;
  
  if (!authKey) {
    return jsonResponse({
      type: 4,
      data: {
        content: 'Auth key is required',
        flags: 64,
      },
    });
  }

  return featureComingSoonResponse('login');
}

/**
 * Handle start-shift command
 */
async function handleStartShiftCommand(interaction, env) {
  return featureComingSoonResponse('start-shift');
}

/**
 * Handle end-shift command
 */
async function handleEndShiftCommand(interaction, env) {
  return featureComingSoonResponse('end-shift');
}

/**
 * Handle shift-status command
 */
async function handleShiftStatusCommand(interaction, env) {
  return featureComingSoonResponse('shift-status');
}

/**
 * Helper function to create JSON responses
 */
function jsonResponse(data, status = 200) {
  return new Response(JSON.stringify(data), {
    status,
    headers: {
      'Content-Type': 'application/json',
    },
  });
}

/**
 * Placeholder response for features not yet implemented in Workers
 */
function featureComingSoonResponse(commandName) {
  return jsonResponse({
    type: 4,
    data: {
      embeds: [
        {
          title: 'Feature in Development',
          description: `The \`${commandName}\` command is available in the traditional bot deployment. Full Cloudflare Workers implementation coming soon!`,
          color: 0xbebefe,
          footer: {
            text: 'Use the traditional bot (bot.py) for full functionality'
          }
        },
      ],
      flags: 64,
    },
  });
}

File: worker.py
"""
Cloudflare Workers entry point for Discord bot using Interactions API.
This allows the bot to run on Cloudflare Workers as a webhook-based application.
"""

import json
import os
from typing import Any, Dict, Optional

from discord_interactions import (
    InteractionType,
    InteractionResponseType,
    verify_key,
)


class DiscordInteractionHandler:
    """Handles Discord interactions for Cloudflare Workers deployment."""
    
    def __init__(self, public_key: str):
        self.public_key = public_key
    
    def verify_signature(self, body: str, signature: str, timestamp: str) -> bool:
        """Verify the Discord interaction signature."""
        try:
            return verify_key(body.encode(), signature, timestamp, self.public_key)
        except Exception:
            return False
    
    def handle_ping(self) -> Dict[str, Any]:
        """Handle PING interaction type."""
        return {
            "type": InteractionResponseType.PONG
        }
    
    def handle_command(self, interaction: Dict[str, Any]) -> Dict[str, Any]:
        """Handle APPLICATION_COMMAND interaction type."""
        command_name = interaction.get("data", {}).get("name")
        
        if command_name == "ping":
            return self._ping_command(interaction)
        elif command_name == "login":
            return self._login_command(interaction)
        elif command_name == "start-shift":
            return self._start_shift_command(interaction)
        elif command_name == "end-shift":
            return self._end_shift_command(interaction)
        elif command_name == "shift-status":
            return self._shift_status_command(interaction)
        else:
            return {
                "type": InteractionResponseType.CHANNEL_MESSAGE_WITH_SOURCE,
                "data": {
                    "content": f"Unknown command: {command_name}",
                    "flags": 64
                }
            }
    
    def _ping_command(self, interaction: Dict[str, Any]) -> Dict[str, Any]:
        """Handle ping command."""
        return {
            "type": InteractionResponseType.CHANNEL_MESSAGE_WITH_SOURCE,
            "data": {
                "embeds": [{
                    "title": "🏓 Pong!",
                    "description": "Bot is running on Cloudflare Workers!",
                    "color": 0xBEBEFE
                }]
            }
        }
    
    def _login_command(self, interaction: Dict[str, Any]) -> Dict[str, Any]:
        """Handle login command - deferred to process async."""
        return {
            "type": InteractionResponseType.DEFERRED_CHANNEL_MESSAGE_WITH_SOURCE,
            "data": {
                "flags": 64
            }
        }
    
    def _start_shift_command(self, interaction: Dict[str, Any]) -> Dict[str, Any]:
        """Handle start-shift command - deferred to process async."""
        return {
            "type": InteractionResponseType.DEFERRED_CHANNEL_MESSAGE_WITH_SOURCE,
            "data": {
                "flags": 64
            }
        }
    
    def _end_shift_command(self, interaction: Dict[str, Any]) -> Dict[str, Any]:
        """Handle end-shift command - deferred to process async."""
        return {
            "type": InteractionResponseType.DEFERRED_CHANNEL_MESSAGE_WITH_SOURCE,
            "data": {
                "flags": 64
            }
        }
    
    def _shift_status_command(self, interaction: Dict[str, Any]) -> Dict[str, Any]:
        """Handle shift-status command - deferred to process async."""
        return {
            "type": InteractionResponseType.DEFERRED_CHANNEL_MESSAGE_WITH_SOURCE,
            "data": {
                "flags": 64
            }
        }
    
    def process_interaction(self, interaction: Dict[str, Any]) -> Dict[str, Any]:
        """Process the interaction and return appropriate response."""
        interaction_type = interaction.get("type")
        
        if interaction_type == InteractionType.PING:
            return self.handle_ping()
        elif interaction_type == InteractionType.APPLICATION_COMMAND:
            return self.handle_command(interaction)
        else:
            return {
                "type": InteractionResponseType.CHANNEL_MESSAGE_WITH_SOURCE,
                "data": {
                    "content": "Unsupported interaction type",
                    "flags": 64
                }
            }


async def handle_request(request):
    """
    Main request handler for Cloudflare Workers.
    
    This function processes incoming Discord interactions.
    """
    public_key = os.getenv("DISCORD_PUBLIC_KEY")
    
    if not public_key:
        return Response(
            json.dumps({"error": "Missing DISCORD_PUBLIC_KEY"}),
            status=500,
            headers={"Content-Type": "application/json"}
        )
    
    if request.method != "POST":
        return Response(
            json.dumps({"error": "Method not allowed"}),
            status=405,
            headers={"Content-Type": "application/json"}
        )
    
    signature = request.headers.get("X-Signature-Ed25519")
    timestamp = request.headers.get("X-Signature-Timestamp")
    
    if not signature or not timestamp:
        return Response(
            json.dumps({"error": "Missing signature headers"}),
            status=401,
            headers={"Content-Type": "application/json"}
        )
    
    body = await request.text()
    
    handler = DiscordInteractionHandler(public_key)
    
    if not handler.verify_signature(body, signature, timestamp):
        return Response(
            json.dumps({"error": "Invalid signature"}),
            status=401,
            headers={"Content-Type": "application/json"}
        )
    
    try:
        interaction = json.loads(body)
        response_data = handler.process_interaction(interaction)
        
        return Response(
            json.dumps(response_data),
            status=200,
            headers={"Content-Type": "application/json"}
        )
    except Exception as e:
        return Response(
            json.dumps({"error": f"Internal error: {str(e)}"}),
            status=500,
            headers={"Content-Type": "application/json"}
        )


class Response:
    """Simple Response class for compatibility."""
    def __init__(self, body, status=200, headers=None):
        self.body = body
        self.status = status
        self.headers = headers or {}
